File: src/mithrillog/storage.py
from __future__ import annotations

import json
from dataclasses import dataclass
from pathlib import Path
from typing import Any, Iterable

from .utils.time import get_timezone, minute_bucket_path


@dataclass
class BucketRecord:
    path: Path
    count: int
    samples: dict[str, list[dict[str, Any]]]


class JournalWriter:
    def __init__(self, base_dir: Path, timezone_name: str = "UTC") -> None:
        self.base_dir = base_dir
        self.base_dir.mkdir(parents=True, exist_ok=True)
        self.bucket_timezone = get_timezone(timezone_name)

    def append(self, event: dict[str, Any], bucket_time: Any = None) -> Path:
        """
        Append event to journal. If bucket_time is provided, use it for bucket path.
        Otherwise, use event timestamp. The event timestamp is always preserved in the record.
        """
        from datetime import datetime
        
        # Use provided bucket_time (server receive time) or fall back to event timestamp
        if bucket_time is None:
            bucket_time = event["timestamp"]
        elif isinstance(bucket_time, datetime):
            pass  # Already a datetime
        else:
            bucket_time = event["timestamp"]  # Fallback
        
        bucket = minute_bucket_path(str(self.base_dir), bucket_time, tz=self.bucket_timezone)
        path = Path(bucket)
        path.parent.mkdir(parents=True, exist_ok=True)
        line = json.dumps(event, separators=(",", ":"), default=str)
        with path.open("a", encoding="utf-8") as handle:
            handle.write(line + "\n")
        return path

    def write_summary(self, path: Path, summary: dict[str, Any]) -> None:
        path.parent.mkdir(parents=True, exist_ok=True)
        with path.open("w", encoding="utf-8") as handle:
            json.dump(summary, handle, indent=2, default=str)

    def write_metadata(self, path: Path, metadata: dict[str, Any]) -> None:
        path.parent.mkdir(parents=True, exist_ok=True)
        with path.open("w", encoding="utf-8") as handle:
            json.dump(metadata, handle, indent=2, default=str)

    def list_buckets(self, start: Path) -> Iterable[Path]:
        for file in sorted(start.rglob("*.ndjson")):
            yield file

File: src/ingester-rs/src/storage.rs
use anyhow::Result;
use chrono::{DateTime, Datelike, Timelike, Utc};
use serde_json::Value;
use std::fs::{self, OpenOptions};
use std::io::Write;
use std::path::{Path, PathBuf};

pub struct JournalWriter {
    base_dir: PathBuf,
}

impl JournalWriter {
    pub fn new(base_dir: PathBuf) -> Self {
        fs::create_dir_all(&base_dir).ok();
        Self { base_dir }
    }

    pub fn bucket_path(&self, time: DateTime<Utc>) -> PathBuf {
        let year = time.year();
        let month = time.month();
        let day = time.day();
        let hour = time.hour();
        let minute = time.minute();

        self.base_dir
            .join(format!("{:04}", year))
            .join(format!("{:02}", month))
            .join(format!("{:02}", day))
            .join(format!("{:02}", hour))
            .join(format!("{:02}.ndjson", minute))
    }

    pub fn append(&self, record: &Value, bucket_time: DateTime<Utc>) -> Result<PathBuf> {
        let path = self.bucket_path(bucket_time);
        
        if let Some(parent) = path.parent() {
            fs::create_dir_all(parent)?;
        }

        let mut file = OpenOptions::new()
            .create(true)
            .append(true)
            .open(&path)?;

        // Use compact JSON format
        let line = serde_json::to_string(record)?;
        writeln!(file, "{}", line)?;

        Ok(path)
    }

    pub fn write_metadata(&self, path: &Path, metadata: &Value) -> Result<()> {
        if let Some(parent) = path.parent() {
            fs::create_dir_all(parent)?;
        }

        let mut file = std::fs::File::create(path)?;
        serde_json::to_writer_pretty(&mut file, metadata)?;
        Ok(())
    }
}
